lib/Backends/OpenCL/kernels.cl

```diff
@@ -1027,50 +1034,147 @@ __kernel void softmaxgradW(__global void *mem, cl_uint32_t origDest,
   softmaxgradK(&mem[srcGrad], &mem[origDest], &mem[selected], sliceSize);
 }
 
+// Aggressive compile-time specialization of convolution parameters.
```

Comment:
The idea of specializing kernels at the source-code level is really nice!

I wonder how to support it in a more principled way, so that we could easily apply the same trick to any of our kernels with a minimal effort. It may require introducing some machinery to support it, e.g. some macros, etc. Do you have any ideas how to expand it in this direction?

Reply:
I tried to think a bit of a more generic specialization mechanism, but couldn't come up with a clean one so far. The problem is the fallback: if we want to partially specialize or not specialize at all using the same source code, the parameters must come from kernel arguments. The current code relies on kernel argument naming. Also since convolutions (and perhaps matmuls) are usually by far the biggest bottlenecks in inference, I didn't yet feel motivated to design a more complex mechanism for this.

And now that I look at it again after your other comment, it might make more sense at least in this simple convolutionK case to not even utilize the same source code for non-specialized and specialized versions for cleanup reasons, but add another source file for a (fully) specialized one and either specialize all paramters or none. Then the end result will look like conv_forward_mem which seems to be always specialized (it refers to it as "parametrized") via macros. I'll look into this.